### PR TITLE
[WebUI] Fix #2798: WebUI plugin fail to start

### PR DESCRIPTION
**Pull request by bendikro**

### CODE REVIEW

File: deluge/ui/web/server.py [hunk 1]
```diff
@@ -567,15 +567,31 @@ def win_handler(ctrl_type):
...
         log.info("%s %s.", _("Starting server in PID"), os.getpid())
+        self.standalone = standalone
         if self.https:
             self.start_ssl()
         else:
             self.start_normal()
 
-        component.get("Web").enable()
-        reactor.run()
+        return component.get("Web").enable()
```
Comment (cas--) [inline deluge/ui/web/server.py]: The return statement here means that the code below is never run...

**Comment by cas--** [comment]
Would it be better to detect if the plugin has been enabled rather than use `standalone` attribute? It may help solve the issue of attempting to start both deluge-web and web plugin.

File: deluge/ui/web/server.py (hunk 1 above)
Comment (bendikro) [inline deluge/ui/web/server.py]: Ah, that extra return should not be there.

**Comment by bendikro** [comment]
What is the problem with both WebUI plugin and deluge-web?

File: deluge/plugins/WebUi/deluge/plugins/webui/__init__.py [hunk 2]
```diff
@@ -26,10 +26,3 @@ def __init__(self, plugin_name):
         from .gtkui import GtkUI as _pluginCls
         self._plugin_cls = _pluginCls
         super(GtkUIPlugin, self).__init__(plugin_name)
-
-
-class WebUIPlugin(PluginInitBase):
```
Comment (cas--) [inline deluge/plugins/WebUi/deluge/plugins/webui/__init__.py]: Leave this code in place

**Comment by cas--** [comment]
In gtk `standalone` mode the webui plugin generates the following errors:

```
12:16:44.400 [ERROR   ][deluge.component                       :98  ] [Failure instance: Traceback: <class 'twisted.internet.error.CannotListenError'>: Couldn't listen on 0.0.0.0:8112: [Errno 98] Address already in use.
/home/ubuntu/deluge.git/deluge/core/pluginmanager.py:56:enable_plugin
/home/ubuntu/deluge.git/deluge/pluginmanagerbase.py:144:enable_plugin
/home/ubuntu/deluge.git/deluge/component.py:286:start
/home/ubuntu/deluge.git/deluge/component.py:104:_component_start
--- <exception caught here> ---
/usr/lib/python2.7/dist-packages/twisted/internet/defer.py:134:maybeDeferred
/home/ubuntu/deluge.git/deluge/plugins/WebUi/deluge/plugins/webui/core.py:74:start
/home/ubuntu/deluge.git/deluge/ui/web/server.py:589:start
/home/ubuntu/deluge.git/deluge/ui/web/server.py:597:start_normal
/usr/lib/python2.7/dist-packages/twisted/internet/posixbase.py:436:listenTCP
/usr/lib/python2.7/dist-packages/twisted/internet/tcp.py:622:startListening
]
12:16:44.410 [ERROR   ][deluge.plugins.init                    :28  ] Unable to enable plugin "GtkPlugin.WebUi"!
12:16:44.410 [ERROR   ][deluge.plugins.init                    :1172] 'NoneType' object has no attribute 'send_request'
Traceback (most recent call last):
  File "/home/ubuntu/deluge.git/deluge/plugins/init.py", line 26, in enable
    self.plugin.enable()
  File "/home/ubuntu/deluge.git/deluge/plugins/WebUi/deluge/plugins/webui/gtkui.py", line 35, in enable
    client.webui.get_config().addCallback(self.cb_get_config)
  File "/home/ubuntu/deluge.git/deluge/ui/client.py", line 517, in __call__
    return self.daemon.call(self.base, *args, **kwargs)
  File "/home/ubuntu/deluge.git/deluge/ui/client.py", line 300, in call
    self.protocol.send_request(request)
AttributeError: 'NoneType' object has no attribute 'send_request'
12:16:44.411 [ERROR   ][deluge.core.eventmanager               :37  ] Event handler PluginEnabledEvent failed in <bound method PluginManager._on_plugin_enabled_event of <deluge.ui.gtkui.pluginmanager.PluginManager object at 0xa40ec6c>> with exception 'NoneType' object has no attribute 'send_request'
12:16:44.414 [ERROR   ][deluge.core.eventmanager               :37  ] Event handler PluginEnabledEvent failed in <bound method JSON.get_remote_methods of <deluge.ui.web.json_api.JSON object at 0xa835dac>> with exception 'NoneType' object has no attribute 'send_request'
12:16:45.094 [ERROR   ][twisted                                :532 ] Unhandled error in Deferred:
Unhandled error in Deferred:
12:16:45.098 [ERROR   ][twisted                                :532 ] Unhandled Error
Traceback (most recent call last):
  File "/usr/lib/python2.7/dist-packages/twisted/internet/_glibbase.py", line 282, in run
    self._run()
  File "/usr/lib/python2.7/dist-packages/twisted/internet/_glibbase.py", line 313, in _simulate
    self.runUntilCurrent()
  File "/usr/lib/python2.7/dist-packages/twisted/internet/base.py", line 800, in runUntilCurrent
    call.func(*call.args, **call.kw)
  File "/usr/lib/python2.7/dist-packages/twisted/internet/task.py", line 208, in __call__
    d = defer.maybeDeferred(self.f, *self.a, **self.kw)
--- <exception caught here> ---
  File "/usr/lib/python2.7/dist-packages/twisted/internet/defer.py", line 134, in maybeDeferred
    result = f(*args, **kw)
  File "/home/ubuntu/deluge.git/deluge/ui/gtkui/statusbar.py", line 367, in update
    self.send_status_request()
  File "/home/ubuntu/deluge.git/deluge/ui/gtkui/statusbar.py", line 271, in send_status_request
    client.core.get_session_status(keys).addCallback(self._on_get_session_status)
  File "/home/ubuntu/deluge.git/deluge/ui/client.py", line 517, in __call__
    return self.daemon.call(self.base, *args, **kwargs)
exceptions.AttributeError: 'NoneType' object has no attribute 'call'
```

**Comment by bendikro** [comment]
The `Couldn't listen on 0.0.0.0:8112: [Errno 98] Address already in use.` indicates the port is taken.

**Comment by bendikro** [comment]
If running deluge-web and GTKUI, and then trying to enable webui plugin I get that error, which is expected, no? [PR68](https://github.com/deluge-torrent/deluge/pull/68) includes improvements to handle this better.

File: deluge/plugins/WebUi/deluge/plugins/webui/__init__.py (hunk 2 above)
Comment (bendikro) [inline deluge/plugins/WebUi/deluge/plugins/webui/__init__.py]: Added back

**Comment by cas--** [comment]
I know that the error indicates that but nothing else is running and the issue doesn't occur in thinclient mode...

**Comment by bendikro** [comment]
Ahh, yeah, that is due to the webui plugin creating two DelugeWeb instances, where the second one causes this trace because the first one is already listening on that port. This is not a problem introduced in this commit, but has existed for a long time. The other PR fixes this.

This commit is not supposed to fix all webui bugs, but simply restore the old behavior (before the latest JSON/WebUI changes)

**Comment by cas--** [comment]
Merged 64c67a0

**Pull request closed without merging**